File: src/browser/runtimeSession.ts
import {
  loadAuthenticatedRuntime,
  rotateAuthenticatedSession,
  saveAuthenticatedRuntime,
  type SessionState,
} from "./authenticatedRuntime.js";
import { computeKaalkaHashPayload } from "../crypto/kaalkaRuntime.js";

export type RuntimeSessionEnvelope = SessionState & {
  session_id: string;
  bounded: boolean;
};

export function createRuntimeSession(state: SessionState = {}): RuntimeSessionEnvelope {
  const normalized: SessionState = {
    cookies: state.cookies ?? [],
    headers: state.headers ?? {},
    auth_tokens: state.auth_tokens ?? [],
    localStorage: state.localStorage ?? {},
    sessionStorage: state.sessionStorage ?? {},
  };
  return {
    ...normalized,
    session_id: computeKaalkaHashPayload(normalized),
    bounded: true,
  };
}

export function persistRuntimeSession(
  path: string,
  session: RuntimeSessionEnvelope,
  encryptionKey: string,
): { path: string; session_id: string; bounded: boolean } {
  const result = saveAuthenticatedRuntime(path, session, encryptionKey);
  return { path: result.path, session_id: session.session_id, bounded: true };
}

export function restoreRuntimeSession(
  path: string,
  encryptionKey: string,
): RuntimeSessionEnvelope {
  const loaded = loadAuthenticatedRuntime(path, encryptionKey);
  return createRuntimeSession(loaded);
}

export function rotateRuntimeSession(session: RuntimeSessionEnvelope): RuntimeSessionEnvelope {
  return createRuntimeSession(rotateAuthenticatedSession(session));
}

File: lib/src/browser/runtime_session.dart
import '../crypto/kaalka_runtime.dart';
import 'authenticated_runtime.dart';

Map<String, dynamic> createRuntimeSession([Map<String, dynamic>? state]) {
  final s = state ?? {};
  final session = {
    'cookies': s['cookies'] ?? <dynamic>[],
    'headers': s['headers'] ?? <String, dynamic>{},
    'auth_tokens': s['auth_tokens'] ?? <dynamic>[],
    'localStorage': s['localStorage'] ?? <String, dynamic>{},
    'sessionStorage': s['sessionStorage'] ?? <String, dynamic>{},
  };
  return {
    ...session,
    'session_id': computeDeterministicHash(session),
    'bounded': true,
  };
}

Map<String, dynamic> persistRuntimeSession(
  String path,
  Map<String, dynamic> session,
  String encryptionKey,
) {
  saveAuthenticatedRuntime(path, session, encryptionKey);
  return {'path': path, 'session_id': session['session_id'], 'bounded': true};
}

Map<String, dynamic> restoreRuntimeSession(String path, String encryptionKey) =>
    createRuntimeSession(loadAuthenticatedRuntime(path, encryptionKey));
